Change autoc[] to double

ktmf01 opened the pull request:
This changes the computation of autocorrelation from single precision to double precision. This results in a compression gain of on average 0.3%point or 0.5% with compression setting -4, but on some tracks this can reach 2.2%point or 5% for
setting -4. See the graphs below

http://www.audiograaf.nl/misc_stuff/double-autoc-with-sse2-intrinsics.pdf
http://www.audiograaf.nl/misc_stuff/double-autoc-with-sse2-intrinsics-per-track.pdf

More information and explanation is on the flac-dev mailing list: http://lists.xiph.org/pipermail/flac-dev/2021-June/006470.html

ktmf01 commented:
The second commit, "Add check for vec_doubleh in altivec.h to configure" should also fix issue #199

ktmf01 commented:
This is driving me crazy. I cannot reproduce the failing CI checks. I've ran the full test suite over 1000 times on my Linux x86 system, (800 times with configure, 200 times with CMake -DENABLE_64_BIT_WORDS), but not fails. I've also borrowed a Macbook from a friend of mine, installed XCode 12.4 and ran CMake -DENABLE_64_BIT_WORDS 3 times, but no fails. Is there any way the CI test could be made a little more verbose?

rillian commented:
The failure is with the macos-latest-clang-cmake-64-bit-words job:

> Operating System
>  - Mac OS X
>  - 10.15.7
>  - 19H1217
>
> Virtual Environment
>  - Environment: macos-10.15
>  - Version: 20210626.1
>  - Included Software: https://github.com/actions/virtual-environments/blob/macOS-10.15/20210626.1/images/macos/macos-10.15-Readme.md
>  - Image Release: https://github.com/actions/virtual-environments/releases/tag/macOS-10.15%2F20210626.1
>
> Build with CMake
>  - cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DENABLE_64_BIT_WORDS=ON -DCMAKE_FIND_FRAMEWORK=NEVER
>  - cmake --build .
>  - ctest -V
>
> 8: noise (--channels=1 --bps=8 -6 -e --lax -b 32768 ): encode... ERROR during encode of noise
> 
> The following tests FAILED:
> 	  8 - streams (Failed)

I don't know if there's a way to make the build logs public.

rillian commented:
Asking github to archive the logs, it appears to be a segfault:
```
### ENCODE noise #######################################################
###    cmd=run_flac --verify --silent --force --force-raw-format --endian=little --sign=signed --sample-rate=44100 --bps=16 --channels=2 -8  --lax -b 65535  --no-padding noise.raw
/Users/runner/work/flac/flac/test/test_streams.sh: line 34: 35087 Segmentation fault: 11  flac${EXE} --no-error-on-compression-fail $*
```

rillian commented:
I spent some time trying to script invoking lldb on failure.

The good new is, if I break the `|| die` on test failure, it fails on about a dozen different variations of the `--channels=2 -8 --lax -b 65535` encoder options on `noise.raw`. Passing `-DENABLE_64_BIT_WORDS=ON` to cmake doesn't make a differnce.

The bad news is, [it doesn't seem to fail under lldb](https://github.com/xiph/flac/runs/2983927674?check_suite_focus=true), so I couldn't get a backtrace.
```
8: noise (--channels=2 --bps=16 -8  --lax -b 65535 ): encode... FAIL
8: (lldb) target create "flac"
8: Current executable set to 'flac' (x86_64).
8: (lldb) run --verify --silent --force --force-raw-format --endian=little --sign=signed --sample-rate=44100 --bps=16 --channels=2 -8  --lax -b 65535  --no-padding noise.raw
8: Process 36323 exited with status = 0 (0x00000000) 
```
Have you tried running the tests under valgrind? I think you have to set `FLAC__TEST_WITH_VALGRIND=yes` explicitly in the environment with the cmake build. The autotools build has an `--enable-valgrind-testing` switch.

ktmf01 commented:
Many, many thanks for taking so much of your time to have an in-depth look into this. I've just started a run with Valgrind, and I see there is indeed something wrong with the new SSE code. Thanks for the tip, I should have thought of that sooner :)

ktmf01 commented:
I've merged and rewritten the additions from pull request #183 (and thereby also #191). Perhaps @coreyjjames might want to take a look at it?

Also, I've done some testing on Travis CI with an altered test_compression.sh (not included in these commits) to check compression and speed, see the very bottom of the logs here: https://travis-ci.com/github/ktmf01/flac/builds/232024699 Results are: VSX code is at least 2x as fast as non-VSX code, compression is exactly the same. NEON code is 30% faster than non-NEON code, compression is exactly the same.

Also, the problem mentioned in #191 should be fixed.

NotTsunami commented on src/libFLAC/lpc_intrin_neon.c:
```diff
@@ -0,0 +1,324 @@
+    vst1q_f64(autoc + 12, sum6);
+    vst1q_f64(autoc + 14, sum7);
+}
+#endif /* FLAC__CPU_ARM && FLAC__HAS_ARCH64INTRIN */
```
```suggestion
#endif /* FLAC__CPU_AARCH64 && FLAC__HAS_ARCH64INTRIN */
```

NotTsunami commented on src/libFLAC/cpu.c:
```diff
@@ -298,6 +300,7 @@ void FLAC__cpu_info (FLAC__CPUInfo *info)
 	case FLAC__CPUINFO_TYPE_PPC:
 		ppc_cpu_info (info);
 		break;
+	case FLAC__CPUINFO_TYPE_ARM: /* fallthrough */
```
```suggestion
	case FLAC__CPUINFO_TYPE_AARCH64: /* fallthrough */
```

NotTsunami commented on src/libFLAC/cpu.c:
```diff
@@ -286,6 +286,8 @@ void FLAC__cpu_info (FLAC__CPUInfo *info)
 	info->type = FLAC__CPUINFO_TYPE_X86_64;
 #elif defined FLAC__CPU_PPC
 	info->type = FLAC__CPUINFO_TYPE_PPC;
+#elif defined FLAC__CPU_ARM
+	info->type = FLAC__CPUINFO_TYPE_ARM;
```
```suggestion
#elif defined FLAC__CPU_AARCH64
	info->type = FLAC__CPUINFO_TYPE_AARCH64;
```

NotTsunami commented on src/libFLAC/include/private/cpu.h:
```diff
@@ -154,6 +154,7 @@ typedef enum {
 	FLAC__CPUINFO_TYPE_IA32,
 	FLAC__CPUINFO_TYPE_X86_64,
 	FLAC__CPUINFO_TYPE_PPC,
+	FLAC__CPUINFO_TYPE_ARM,
```
```suggestion
	FLAC__CPUINFO_TYPE_AARCH64,
```

NotTsunami reviewed (changes requested):
At some point `FLAC__CPU_ARM` was changed to `FLAC__CPU_AARCH64` in the old PR, and there are still a few remnants of that here. I'll probably throw out another review once I take a more in-depth look.

NotTsunami commented:
None of those changes are functional changes (in the above review), but it'll keep naming consistent.

ktmf01 commented:
I've just taken a look myself, I think those are indeed all mentions of 'ARM' in the code. However, now that I think of it, wouldn't it be better to completely remove the ARM-related code in cpu.c, as it does not serve any purpose right now? Or should it be kept around in case anyone wants to write code with SVE/SVE2 intrinsics? What do you think @NotTsunami ?

NotTsunami commented:
> I've just taken a look myself, I think those are indeed all mentions of 'ARM' in the code. However, now that I think of it, wouldn't it be better to completely remove the ARM-related code in cpu.c, as it does not serve any purpose right now? Or should it be kept around in case anyone wants to write code with SVE/SVE2 intrinsics? What do you think @NotTsunami ?

It would be nice to have some ARMv7 compatible NEON intrinsics, but since this switches autoc[] to double-precision, it would be more of a hassle for that. For simplicity's sake, you could probably remove it for now, and if anyone picks up any further intrinsics, they could probably add it back in.

ktmf01 commented:
I've removed the ARM parts of cpu.c. As NEON is optional in ARMv7 and requires OS-specific code to check for support (unlike x86 feature detection), this would need quite a bit of rework anyway.

ktmf01 commented:
Sorry I keep adding commits to this pull request, but I saw quite an opportunity further optimizing the code I already touched. I rewrote the VSX and NEON code to be much cleaner and easier to read, and added a check to the VSX code for very small data_len. This gives a speed increase for preset 8 of about 40%.

I also rewrote the SSE2 code to be much cleaner and easier to read, but sadly the gain is only 8%.

ktmf01 commented:
Earlier it was mentioned that this change results in on average 0.5% compression improvement, with outliers on specific material of up to 5%, but it turns out the gains are much, much larger on files where high-frequency signals are missing, for example band-passed files, 78rpm vinyl rips or high samplerate (>96kHz) material that contains only audible frequencies. As noted in https://hydrogenaud.io/index.php?topic=120158.msg1001482#msg1001482, gains of 10.7%-point (or 19.5%) have been seen.

ValZapod commented:
Two questions: afaik, 32 bit audio with all bits set can be put into 24 bit flac losslessly. But what about adding actual support for 32-bit flac? Everybody says there are no implementations of 32 bit flac even here in the reference encoder.

Second: I see that in the link ahove (https://hydrogenaud.io/index.php?topic=120158.msg1001482#msg1001482) somebody brute-forces flac for 5 days to produce even smaller samples. Did you think about creating an algorithm that will converge faster or an AI approach with then reverse ingeenering the algorithm?

ktmf01 commented:
@ValZapod The two question you are asking have nothing to do with this pull request. Concerning the 32-bit audio, you could open a new issue, but please provide a clear use case. Concerning the second question, please read the first few posts of that thread carefully. You'll see that there are still people trying new algorithms, but these developments take **a lot** of time. Please keep discussion in this thread about this pull request.

ktmf01 commented:
@antonblanchard I haven't been able to contact you by mail, so I try it over github: Could you perhaps spare some time to check the changes this pull request makes to the PPC code you added a few years back? The performance claims I made concerning the speedups were based on limited data supplied by the CI running on PowerPC, but perhaps you can verify those too?

ktmf01 commented:
I'm going to split up this PR because it does way to much at once. I will split off the ARM64 part and rework the SSE2/PPC parts

vrubleg commented:
Please mention this PR in the new PRs, so everybody who are following it would know what to follow now. I hope that this cool thing will be merged soon :)

ktmf01 commented:
@vrubleg most is now in #317 and #318. For the ARM intrinsics part, that still needs some more work.
